File: Two_Pointers/015.3Sum/015.3Sum.ts
function threeSum(nums: number[]): number[][] {
  let ans: number[][] = [];
  nums.sort((a, b) => a - b);
  let n = nums.length;
  for (let i = 0; i < n; i++) {
    let left = i + 1;
    let right = n - 1;
    const sum = -nums[i];
    while (left < right) {
      if (nums[left] + nums[right] === sum) {
        ans.push([nums[i], nums[left], nums[right]]);
        left++;
        right--;
        while (left < right && nums[left] === nums[left - 1]) {
          left++;
        }
        while (left < right && nums[right] === nums[right + 1]) {
          right--;
        }
      } else if (nums[left] + nums[right] > sum) {
        right -= 1;
      } else {
        left += 1;
      }
    }
    while (i + 1 < n && nums[i + 1] === nums[i]) {
      i++;
    }
  }
  return ans;
}

File: Two_Pointers/015.3Sum/015.3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums);
        result = [];

        for i in range(len(nums)):
            if (i > 0 and nums[i] == nums[i - 1]):
                continue;

            left = i + 1;
            right = len(nums) - 1;
            target = -nums[i];

            while left < right:
                if nums[left] + nums[right] == target:
                    result.append([nums[i], nums[left], nums[right]]);
                    left += 1;
                    right -=1;

                    while left < right and nums[left] == nums[left - 1]:
                        left += 1;
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1;
                elif nums[left] + nums[right] > target:
                    right -= 1;
                else:
                    left += 1;
            
            while i + 1 < len(nums) and nums[i] == nums[i + 1]:
                i += 1;

        return result;
